### PR TITLE
Don't use fibers as keys for native Map and Set

### PR DESCRIPTION
**Pull request by acdlite**
Most Map and Set polyfills work by adding a hidden field to the key. We don't want to do this for fibers because it causes a de-opt. Because the only case where we use fibers as keys is to keep track of failed error boundaries, and the number of failed error boundaries is usually low, we can use arrays instead.

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberCollection.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberCollection
```
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberCollection.js]: Can we put "Linear" or "Array" or something in the name to remind people of perf characteristics?

File: src/renderers/shared/fiber/ReactFiberCollection.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      return true;
+    }
+  }
+  return false;
+}
+exports.fiberMapHas = fiberMapHas;
+
+function fiberMapSet<T>(map: FiberMap<T>, fiber: Fiber, value: T) {
+  if (fiberMapHas(map, fiber)) {
+    return;
```
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberCollection.js]: If we're not gonna support this properly (by overwriting the value), can we throw?

**Comment by acdlite** [comment]
> If FiberSet is intentionally a leaky abstraction (transparently an Array)

Because of the `forEach`? Leakiness not intentional, I missed that because both `Set` and `Array` happen to have that method. Generally I'd prefer people don't treat FiberSet as an array in case we want to swap the implementation again.

**Comment by acdlite** [comment]
Talked to @sebmarkbage and we decided this is too much code to be worth it. We have an alternate strategy, will open a separate PR.

**Pull request closed without merging**